### PR TITLE
Fix python implementation

### LINKED ISSUE #155: `xb.get_bitinformation(implementation="julia/python")` yields different results
* xbitinfo version: main
* Python version: binder
* Operating System: binder

### Description

- @observingClouds shouldnt these yield same results?
- I am lacking a test in #126 in https://github.com/observingClouds/xbitinfo/blob/main/tests/test_get_bitinformation.py where the numerical values of both implementations are compared and are equal
- makes https://xbitinfo--153.org.readthedocs.build/en/153/quick-start.html fail

### What I Did

```python
ds = xr.tutorial.load_dataset("eraint_uvz")
xb.get_bitinformation(ds, dim="longitude", implementation="python")
xb.get_bitinformation(ds, dim="longitude", implementation="julia")
```

### PR DESCRIPTION
**Pull request by observingClouds**
fixes #155

### CODE REVIEW

File: tests/test_get_bitinformation.py [hunk 1]
```diff
@@ -215,3 +215,23 @@ def test_get_bitinformation_different_dtypes(rasm, implementation):
 def test_get_bitinformation_dim_list(rasm, implementation):
     bi = xb.get_bitinformation(rasm, dim=["x", "y"], implementation=implementation)
     assert (bi.dim == ["x", "y"]).all()
+
+
+def test_implementations_agree():
+    """Test whether the python and julia implementation retrieve the same results"""
+    ds = xr.tutorial.load_dataset("rasm")
```
Comment (aaronspring) [inline tests/test_get_bitinformation.py:222]: Does this also work for other tutorial datasets?

**Review by aaronspring — APPROVED** [review]

File: tests/test_get_bitinformation.py (hunk 1 above)
Comment (observingClouds) [reply tests/test_get_bitinformation.py:222]: It should! I tested it locally with the `era5int_uvz`. Are you suggesting, we should test for several datasets here?

Comment (aaronspring) [reply tests/test_get_bitinformation.py:222]: I think I parametrized the input datasets in some other tests

Comment (aaronspring) [reply tests/test_get_bitinformation.py:222]: parametrize: dataset name, dim

**Comment by rsignell-usgs** [comment]
So the python and julia versions don't agree because of different feature sets, right?

**Comment by observingClouds** [comment]
Currently I think that there is still a small bug in the python implementation. Something with the dimension handling doesn't seem right.

**Comment by rsignell-usgs** [comment]
I'm eagerly awaiting the stamp of approval the python method.  Is this something we might be able to hire someone from a company (e.g. Anaconda or Quansight) to investigate and fix, or does it require expert knowledge that would make that approach impractical?

**Comment by observingClouds** [comment]
@rsignell-usgs I made a mistake somewhere in the algorithm which I just need to find. Your idea seems great, thank you for the offer, but because this issue is within the algorithm itself one needs to familiarise with that. I haven't found time to look at this again.

**Comment by milankl** [comment]
Happy to be of help here if there's anything I can do. I hope the Julia algorithm is pretty clear (let me know if not!). These might be naive questions and you *can ignore them*, but maybe they also help to find the bug. I struggle to follow the python implementation because

- you only seem to check for the masked values after the bitpair_count runs. I don't know how this should be possible, because if either value is masked neither bit should be counted in the first place. As far as I understand this line `index = ((unpack_a << 1) | unpack_b).reshape(-1, 8)` both `unpack_a` and `_b` are matrices of with every entry either 0x00 or 0x01 hence you shift `_a` by `<< 1` to get the bitpairs. The `|` operator then turns every element in the resulting matrix into 0b0000_0000, ..01, ..10, or ..11. While I get this logic I don't see how you can incorporate the check for masking after the reduction operation which seems to happen here `to_return = sel.sum(axis=0).reshape(8, 2, 2)` 
- the loop seems to be over bytes of the input arrays which would lead to a lot of non-contiguous memory access? Also `unpack_a` and `_b` are matrices of type uint8 for every bit in the origional array `a`. So if `a` is 1 GB in Float32, then you loop over 250MB chunks (per byte), but using a uint8 for every bit would blow this up to 2*8*250MB=4GB? Not speaking of the `|` operator which presumably also allocates another one, so 6GB? Maybe I'm naive and don't understand how python handles memory allocations though!!
- you seem to apply bitwise operators to whole arrays, which I don't how python handles that, but in my naive understanding either lead to massive amounts of memory allocations, or to undesired in-place operations. like
```python
for i in range(nbytes):
        s = (nbytes - 1 - i) * 8
        bitc = bitpaircount_u1((a >> s).astype("u1"), (b >> s).astype("u1"))
```
here either you create a copy of `a,b` which is bitshifted for every element, or it happens in-place in which case you probably want to shift every iteration by 8 not by ...24,16,8,0?

**Comment by rsignell-usgs** [comment]
Thanks @milankl for taking a look!

**Comment by observingClouds** [comment]
You are both too fast...or I'm too slow (probably the latter). Let's see if I can find some time next week to look at this again!

**Comment by rsignell-usgs** [comment]
@observingClouds , thanks for trying to carve out some time!   I think the impact will be huge, as most of the modelers I know are using python!

**Comment by rsignell-usgs** [comment]
> Sure, in the ideal world we would just need to wrap your code, but so far the wrapping of Julia code introduces a lot of memory duplications and doesn't work well with dask

@observingClouds I was going to say exactly that!   :)

**Comment by milankl** [comment]
> We will see how the performance will compare, but I'm positive it will be alright.

For comparison:
```julia
julia> A = rand(Float32,2_500_000);    # 10MB array
julia> @btime bitinformation($A);
  123.739 ms (279 allocations: 13.19 KiB)
```
that's about 80MB/s (single-threaded). The allocations are the counter and the returned array, so they do not scale with the size of the input array. If you use masked arrays, the performance goes up proportional to masking:
```julia
julia> mask = A .< 0.1    # mask 10%
julia> @btime bitinformation($A,$mask);
  108.613 ms (273 allocations: 13.12 KiB)
```
which is about 10% faster with ~90MB/s

**Comment by observingClouds** [comment]
Alright, I had a closer look at the implementation discrepancy today and they only seem to occur in the exponent bits. Sign-bit and mantissa-bits seem to be consistent. Now I have to figure out why...
@milankl is there something specific you do to exponent bits? Maybe the order of the exponent bits is not correct yet.

**Comment by observingClouds** [comment]
Interestingly though, for some datasets the returned xbitinfo is identical.

**Comment by observingClouds** [comment]
```python
<xarray.Dataset>
Dimensions:       (time: 1, nMesh2_face: 113885)
Coordinates:
  * time          (time) datetime64[ns] 1950-01-01T04:11:59.999742507
Dimensions without coordinates: nMesh2_face
Data variables:
    Mesh2_height  (time, nMesh2_face) float32 4.069e+03 3.69e+03 ... 4.117e+03
    Mesh2_bathy   (time, nMesh2_face) float32 6.931e+03 7.31e+03 ... 6.883e+03
    Mesh2_m_x     (time, nMesh2_face) float32 -3.357e-10 ... -3.056e-10
    Mesh2_m_y     (time, nMesh2_face) float32 1.091e-10 -9.543e-11 ... -3.47e-10
    Mesh2_u_x     (time, nMesh2_face) float32 -8.25e-14 ... -7.422e-14
    Mesh2_u_y     (time, nMesh2_face) float32 2.681e-14 ... -8.429e-14
Attributes:
    title:              netCDF output from StormFlash2d
    institution:        University of Hamburg, KlimaCampus
    contact:            None
    source:             None
    references:         None
    comment:            None
    Conventions:        UGRID-0.9
    creation_date:      2014-10-03 15:29:40  02:00
    modification_date:  2014-10-03 15:29:40  02:00
```
For `Mesh2_u_x` and `Mesh2_u_y` both implementations return identical results (except for some 1e-17 differences). All other variables of this dataset have differences in the exponent bits (e1-e7).

**Comment by milankl** [comment]
> Alright, I had a closer look at the implementation discrepancy today and they only seem to occur in the exponent bits. Sign-bit and mantissa-bits seem to be consistent. Now I have to figure out why... maybe the order of the exponent bits is not correct yet. @milankl is there something specific you do to exponent bits?

I don't by default, but you have implemented the bitinformation with an prepended conversion to a signed exponent

https://github.com/observingClouds/xbitinfo/blob/72348e9c20c3d36edd8e7664b756e91111f88bf0/xbitinfo/bitinformation_wrapper.jl#L6-L7

For context, the IEEE standard encodes a biased exponent so that not 2^00000000 is 1 but something really tiny, because it's actually 2^(00000000-bias); then further powers of two by counting up till 2^(01111111-bias) is 1, 2^(1000000-bias) is 2 and so on. By this definition every time you jump between <2 and >=2 not just one expont bit flips but all 8 of them. See below
```julia
julia> using BitInformation
julia> x = [2f0^i for i in -5:5]
julia> xs = signed_exponent(x)
julia> cat(x,bitstring.(x,:split),bitstring.(xs,:split),dims=2)
11×3 Matrix{Any}:
  0.03125  "0 01111010 00000000000000000000000"  "0 10000101 00000000000000000000000"
  0.0625   "0 01111011 00000000000000000000000"  "0 10000100 00000000000000000000000"
  0.125    "0 01111100 00000000000000000000000"  "0 10000011 00000000000000000000000"
  0.25     "0 01111101 00000000000000000000000"  "0 10000010 00000000000000000000000"
  0.5      "0 01111110 00000000000000000000000"  "0 10000001 00000000000000000000000"
  1.0      "0 01111111 00000000000000000000000"  "0 00000000 00000000000000000000000"
  2.0      "0 10000000 00000000000000000000000"  "0 00000001 00000000000000000000000"
  4.0      "0 10000001 00000000000000000000000"  "0 00000010 00000000000000000000000"
  8.0      "0 10000010 00000000000000000000000"  "0 00000011 00000000000000000000000"
 16.0      "0 10000011 00000000000000000000000"  "0 00000100 00000000000000000000000"
 32.0      "0 10000100 00000000000000000000000"  "0 00000101 00000000000000000000000"
```
(value | bitstring for biased exponent | bitstring for signed exponent)

That means for a dataset with values in [1,4) you'll always get the same bitinformation for all the exponent bits. This can be easily interpreted as "there's information in all the exponent bits!!!", and in fact, there is, but it's redundant across exponent bits. It's an artefact of the encoding just because it so happens that the "8 bits flip simultaneously" situation is located somewhat inconveniently (for us) around the rather common value 2.

But there are other ways to encode the exponent, see [here](https://en.wikipedia.org/wiki/Signed_number_representations#Comparison_table) for an overview of various ways to encode signed integers (because that's what it is). I wrote the `signed_exponent` function to circumvent this easily misleading artefact as a signed exponent moves the "8 bits flip simultaneously" situation out of the way.

> Interestingly though, for some datasets the returned bitinfo is identical.

That's the case when all data values are (-1,1) as you can see from above that the exponent bits for biased vs signed are the negation of another, so entropy/information is identical.

**TL;DR**
- signed/biased exponent bitinformation are both correct but the latter has a tendency to include misleading redundant bitinformation across exponent bits. So if possible I use a signed exponent.
- for data in (-1,1) signed and biased exponents give the exact same bitinformation

**Comment by observingClouds** [comment]
@milankl, thanks for the detailed explanation. I guess we have found the bug (or rather inconsistency)! Great teamwork! So I'll write a `signed_exponent` function in python and then the basic bitinformation functionality should be ported to python 🤞

**Comment by milankl** [comment]
Do you currently also plan to support masked arrays with your python implementation? Just because, regarding my questions above, I don't see how this is correctly implemented, but maybe I'm misinterpreting the python code from just reading over it.

**Comment by observingClouds** [comment]
Masked arrays won't be supported for now, that is correct. I'm happy if the most basic implementation works. This will hopefully allow others to add features more easily as well.

**Comment by milankl** [comment]
I've tested `signed_exponent` against its inverse `biased_exponent` as I wanted this bit transformation to be reversible. I understand it if you don't want to implement a version of `biased_exponent` too, but for testing this might be helpful
```julia
julia> using BitInformation

julia> powers_of_2 = [2f0^(i-Base.exponent_bias(Float32)) for i in 1:2^8-1];
julia> all_exponents = vcat(0f0,powers_of_2)
256-element Vector{Float32}:
  0.0
  1.1754944f-38
  2.3509887f-38
  4.7019774f-38
  9.403955f-38
  1.880791f-37
  ⋮
  2.1267648f37
  4.2535296f37
  8.507059f37
  1.7014118f38
 Inf

julia> all_exponents_signed = signed_exponent(all_exponents);
julia> cat(all_exponents,bitstring.(all_exponents,:split),all_exponents_signed,bitstring.(all_exponents_signed,:split),dims=2)
256×4 Matrix{Any}:
  0.0          "0 00000000 00000000000000000000000"  Inf          "0 11111111 00000000000000000000000"
  1.17549f-38  "0 00000001 00000000000000000000000"  1.70141f38   "0 11111110 00000000000000000000000"
  2.35099f-38  "0 00000010 00000000000000000000000"  8.50706f37   "0 11111101 00000000000000000000000"
  4.70198f-38  "0 00000011 00000000000000000000000"  4.25353f37   "0 11111100 00000000000000000000000"
  9.40395f-38  "0 00000100 00000000000000000000000"  2.12676f37   "0 11111011 00000000000000000000000"
  1.88079f-37  "0 00000101 00000000000000000000000"  1.06338f37   "0 11111010 00000000000000000000000"
...
  0.25         "0 01111101 00000000000000000000000"  8.0          "0 10000010 00000000000000000000000"
  0.5          "0 01111110 00000000000000000000000"  4.0          "0 10000001 00000000000000000000000"
  1.0          "0 01111111 00000000000000000000000"  0.0          "0 00000000 00000000000000000000000"
  2.0          "0 10000000 00000000000000000000000"  1.17549f-38  "0 00000001 00000000000000000000000"
  4.0          "0 10000001 00000000000000000000000"  2.35099f-38  "0 00000010 00000000000000000000000"
...
  8.50706f37   "0 11111101 00000000000000000000000"  0.5          "0 01111110 00000000000000000000000"
  1.70141f38   "0 11111110 00000000000000000000000"  1.0          "0 01111111 00000000000000000000000"
 Inf           "0 11111111 00000000000000000000000"  2.0          "0 10000000 00000000000000000000000"
```

**Comment by observingClouds** [comment]
Well, I guess this is it 🚀 . I will merge this PR for now as this was mainly for fixing the inconsistency between the julia and python implementation. I'm sure there are still some performance or scaling issues, but those can be discussed in new issues.

**Comment by rsignell-usgs** [comment]
@observingClouds any chance of a 0.0.4 release, now that we have this new capability?  This would make it easier for conda users to install...

**Comment by ocefpaf** [comment]
> This would make it easier for conda users to install...

Actually all users, pip included. A dev install of a package that is always an extra barrier that most folks struggle with.
